### PR TITLE
Support to break wrap the text block when triggering shift+enter

### PR DESCRIPTION
**Pull request by qinluhe**
### Feature Preview

- [x] Refactor: TextBlock's onKeyDown code 
- [x]  Support to break wrap the text block when triggering `shift+enter`

---

#### PR Checklist

- [ ] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [ ] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

### CODE REVIEW

File: frontend/appflowy_tauri/src/appflowy_app/components/document/TextBlock/TextBlock.hooks.ts [hunk 1]
```diff
@@ -99,11 +48,89 @@ export function useTextBlock(id: string) {
...
+      event.preventDefault();
+      matchKey.handler(event, editor);
+    },
+    [editor, enterEvent, backSpaceEvent, tabEvent]
+  );
+
+  return {
+    onKeyDown,
+  };
+}
```
Comment (LucasXu0) [inline frontend/appflowy_tauri/src/appflowy_app/components/document/TextBlock/TextBlock.hooks.ts:128]: missing a new line after `}`

File: frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts [hunk 2]
```diff
@@ -14,9 +16,72 @@ const HOTKEYS: Record<string, string> = {
...
+    void onWrap([...retain, { insert: '\n' }, ...insert], selection);
+    return;
+  }
+
+  // retain this node and insert a new node
+  void onSplit(retain, insert);
+}
+
+function getSelectionAfterBreakWrap(editor: Editor) {
+  const start = Range.start(editor.selection!);
```
Comment (LucasXu0) [inline frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts:80]: try not to use `!`.

File: frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts [hunk 3]
```diff
@@ -14,9 +16,72 @@ const HOTKEYS: Record<string, string> = {
 export function triggerHotkey(event: React.KeyboardEvent<HTMLDivElement>, editor: Editor) {
   for (const hotkey in HOTKEYS) {
     if (isHotkey(hotkey, event)) {
```
Comment (LucasXu0) [inline frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts:18]: same as the previous PR' comment.

I prefer using find to reduce the number of {} in the code.

```ts
const hotkey = HOTKEYS.find(...);
if (hotkey) {
  // ...
}
```

File: frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts [hunk 4]
```diff
@@ -14,9 +16,72 @@ const HOTKEYS: Record<string, string> = {
...
-}
\ No newline at end of file
+}
+
+export function canHandleEnterKey(event: React.KeyboardEvent<HTMLDivElement>, editor: Editor) {
+  const isEnter = event.key === 'Enter';
+  if (!isEnter || !editor.selection) {
+    return false;
+  }
+  return true;
```
Comment (LucasXu0) [inline frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts:31]: Is it equal to `return event.key == 'enter' && editor.selection != null`

File: frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts [hunk 5]
```diff
@@ -14,9 +16,72 @@ const HOTKEYS: Record<string, string> = {
...
+
+export function canHandleBackspaceKey(event: React.KeyboardEvent<HTMLDivElement>, editor: Editor) {
+  const isBackspaceKey = event.key === 'Backspace';
+  if (!isBackspaceKey || !editor.selection) {
+    return false;
+  }
+  // It should be handled if the selection is collapsed and the cursor is at the beginning of the block
+  const { anchor } = editor.selection;
+  const isCollapsed = Range.isCollapsed(editor.selection);
+  return isCollapsed && anchor.offset === 0 && anchor.path.toString() === '0,0';
```
Comment (LucasXu0) [inline frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts:42]: use `editor.selection` many times here. how about assign it to a local value.

```ts
const { selection } = editor;
if (!isBackspaceKey || selection == null) {
  return false;
}
const { anchor } = selection;
// ....
```

I'm not sure the grammar is correct.

File: frontend/appflowy_tauri/src/appflowy_app/utils/slate/text.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    focus: start,
+  };
+}
+
+export function getInsertRangeBy(editor: Editor) {
+  const end = Editor.end(editor, editor.selection!);
+  const fragment = (editor.children[0] as Element).children;
+  return {
+    anchor: end,
+    focus: { path: [0, fragment.length - 1], offset: (fragment[fragment.length - 1] as Text).text.length },
```
Comment (LucasXu0) [inline frontend/appflowy_tauri/src/appflowy_app/utils/slate/text.ts:28]: `(fragment[fragment.length - 1] as Text).text.length`

This line seems a bit complex.

```ts
const last = fragment[fragment.length - 1] // By the way, it looks like the fragment variable is an array or a list. Can we get the last value of the array using `fragment.last`?
const offset = last ? last.text.length : 0; 
```

File: frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts (hunk 3 above)
Comment (qinluhe) [reply frontend/appflowy_tauri/src/appflowy_app/utils/slate/hotkey.ts:18]: It is object, not array

File: frontend/appflowy_tauri/src/appflowy_app/utils/slate/text.ts (hunk 6 above)
Comment (qinluhe) [reply frontend/appflowy_tauri/src/appflowy_app/utils/slate/text.ts:28]: Maybe we just do it:
const lastIndex = fragment.length - 1;
//...
focus: { path: [0, lastIndex], offset: (fragment[lastIndex] as Text).text.length },

**Comment by LucasXu0** [comment]
LGTM.

**Pull request merged**